File: app/src/main/java/com/cs407/team15/redstone/ui/profile/RecyclerAdapter.java
package com.cs407.team15.redstone.ui.profile;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.RecyclerView;

import com.cs407.team15.redstone.MainActivity;
import com.cs407.team15.redstone.R;
import com.cs407.team15.redstone.model.Comment;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import com.cs407.team15.redstone.model.Tour;

import java.util.List;

public class RecyclerAdapter extends RecyclerView.Adapter<RecyclerAdapter.ViewHolder> {

    private List<Tour> mData;
    private List<String> tourIDList;
    private LayoutInflater mInflater;
    private Context context;
    private ItemClickListener mClickListener;
    private ArrayList<Comment> comments;
    private  UserToursFragment userTourFrag;

    // data is passed into the constructor
    RecyclerAdapter(Context context, List<Tour> data,List<String> tourIDList, UserToursFragment userToursFrag) {
        this.mInflater = LayoutInflater.from(context);
        this.context=context;
        this.mData = data;
        this.userTourFrag = userToursFrag;
        this.tourIDList=tourIDList;
    }
    RecyclerAdapter(Context context, List<String> data) {
        this.mInflater = LayoutInflater.from(context);
        this.context=context;
        this.tourIDList=data;
    }

    RecyclerAdapter(Context context, List<String> data, ArrayList<Comment> comments) {
        this.mInflater = LayoutInflater.from(context);
        //this.mData = data;

        this.tourIDList=data;
        this.comments = comments;
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.recycler_row_uc, parent, false);
        //for user tour list page, change text of button
        Button btn = view.findViewById(R.id.delete_tag_button);
        if (comments == null) {
            btn = view.findViewById(R.id.delete_tag_button);
            btn.setText("DELETE TOUR");
        }
        else{
            btn.setText("Delete Tag");
        }
        return new ViewHolder(view);
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, final int position) {
        if(userTourFrag!=null){
        String animal = mData.get(position).getName();
        holder.myTextView.setText(animal);
            /*holder.editTourButton.setOnClickListener(new View.OnClickListener() {
                public void onClick(View view) {
                    mClickListener.onEditClick(mData.get(position),tourIDList.get(position));
                }
            });*/

        }else{
            String animal = tourIDList.get(position);
            holder.myTextView.setText(animal);
        }
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return tourIDList.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView myTextView;
        Button delButton;
        ViewHolder(final View itemView) {
            super(itemView);
            myTextView = itemView.findViewById(R.id.tvAnimalName);
            delButton = itemView.findViewById(R.id.delete_tag_button);
            itemView.setOnClickListener(this);
            delButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (delButton.getText().equals("DELETE TOUR")) {
                        final FirebaseFirestore db = FirebaseFirestore.getInstance(); //watch this
                        final View v2 = v;
                        db.collection("tours")
                                .whereEqualTo("user_id", FirebaseAuth.getInstance().getCurrentUser().getUid())
                                .whereEqualTo("name", mData.get(getAdapterPosition())).get().addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                            @Override
                            public void onComplete(@NonNull Task<QuerySnapshot> task) {
                                if (task.isSuccessful()) {
                                    //ArrayList<String> list = new ArrayList<>();
                                    for (QueryDocumentSnapshot document : task.getResult()) {
                                        db.collection("tours").document(document.getId()).delete().addOnSuccessListener(new OnSuccessListener<Void>() {
                                            @Override
                                            public void onSuccess(Void aVoid) {
                                                Toast toast = Toast.makeText(v2.getContext(), "Tour deleted", Toast.LENGTH_SHORT);
                                                toast.show();
                                                mData.remove(getAdapterPosition());
                                                notifyItemChanged(getAdapterPosition());
                                                //RecyclerAdapter.this.notifyItemChanged(getAdapterPosition());
                                            }
                                        });
                                    }
                                }
                            }
                        });
                    }
                    else{
                    DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Comments").child(comments.get(getAdapterPosition()).getPath()).child(comments.get(getAdapterPosition()).getLocationId())
                            .child(comments.get(getAdapterPosition()).getCommentid()).child("tags");

                    reference.removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            Toast toast = Toast.makeText(itemView.getContext(), "Removed Tag", Toast.LENGTH_SHORT);
                            toast.show();
                            //notifyItemChanged(getAdapterPosition());
                        }
                    });
                }
                }
            });
        }

        @Override
        public void onClick(View view) {
            if (mClickListener != null) mClickListener.onItemClick(view, getAdapterPosition());
        }
    }

    // convenience method for getting data at click position
    String getItem(int id) {
        return tourIDList.get(id);
    }

    // allows clicks events to be caught
    void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, int position);
        void onEditClick(Tour tour , String tourId);
    }
}

File: app/src/main/java/com/cs407/team15/redstone/ui/viewtours/RecyclerAdapter.kt
package com.cs407.team15.redstone.ui.viewtours

import android.annotation.TargetApi
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.cs407.team15.redstone.R
import com.google.firebase.firestore.FirebaseFirestore

class RecyclerAdapter
// data is passed into the constructor
internal constructor(context: Context, private val mData: List<String>) :
    RecyclerView.Adapter<RecyclerAdapter.ViewHolder>() {
    private val mInflater: LayoutInflater
    private var mClickListener: ItemClickListener? = null

    init {
        this.mInflater = LayoutInflater.from(context)
    }

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = mInflater.inflate(R.layout.recycler_row, parent, false)
        return ViewHolder(view)
    }

    // binds the data to the TextView in each row
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val animal = mData[position]
        addHammerIcon(position, holder)
        addTourMileage(position, holder)
        holder.myTextView.text = animal
    }
    @TargetApi(24)
    internal fun addHammerIcon(position: Int, holder: ViewHolder) {
        FirebaseFirestore.getInstance().collection("tours").whereEqualTo("name", mData[position])
            .get().addOnSuccessListener { documents ->
                for (document in documents) {
                    FirebaseFirestore.getInstance().collection("users").whereEqualTo("uid", document["user_id"]).get()
                        .addOnSuccessListener { documents ->
                            for (document in documents) {
                                val one = 1

                                val two = Math.toIntExact(document["userType"] as Long)
                                if (one == two) {
                                    holder.myImageView.visibility = View.VISIBLE

                                }
                            }
                        }
                }
            }
            .addOnFailureListener {
                exception ->
                System.out.println("ERROR")
            }
    }

    internal fun addTourMileage(position: Int, holder: ViewHolder) {
        FirebaseFirestore.getInstance().collection("tours").whereEqualTo("name", mData[position])
            .get().addOnSuccessListener { documents ->
                for (document in documents) {
                    var tourTimeStr = ""
                    val tourTime: Int
                    if (document["time"] != null) {
                        tourTime = (document["time"] as Long).toInt()
                        if (tourTime >= 60) {
                            tourTimeStr =
                                (tourTime / 60).toString() + "h " + (tourTime % 60).toString() + "m"
                        } else if (tourTime < 60 && tourTime >= 0) {
                            tourTimeStr = tourTime.toString() + " min"
                        }
                        holder.myMilesView.text =
                            (tourTimeStr + " " + document["distance"].toString() + " mi")
                    } else {
                        // should not come here but watever TODO fix`
                    }
                }
            }
            .addOnFailureListener {
                    exception ->
                System.out.println("ERROR")
            }
    }

    // total number of rows
    override fun getItemCount(): Int {
        return mData.size
    }


    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        internal var myTextView : TextView
        internal var myImageView: ImageView
        internal var myMilesView : TextView

        init {
            myTextView = itemView.findViewById(R.id.tvAnimalName)
            myImageView = itemView.findViewById(R.id.hammertouricon)
            myMilesView = itemView.findViewById(R.id.viewTourMileage)
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            if (mClickListener != null) mClickListener!!.onItemClick(view, adapterPosition)
        }
    }

    // convenience method for getting data at click position
    internal fun getItem(id: Int): String {
        return mData[id]
    }

    // allows clicks events to be caught
    internal fun setClickListener(itemClickListener: ItemClickListener) {
        this.mClickListener = itemClickListener
    }

    // parent activity will implement this method to respond to click events
    interface ItemClickListener {
        fun onItemClick(view: View, position: Int)
    }
}
